include/deal.II/base/lazy.h

```diff
@@ -0,0 +1,215 @@
+ *   {
+ *     prolongation_matrix.initialize([&](){
+ *       // ... Some expensive operation initializing the prolongation
+ *       // matrix that we only want to perform once and when necessary.
+ *       });
+ *     return prolongation_matrix;
+ *   }
+ *
+ * private:
+ *   mutable Lazy<FullMatrix<double>> prolongation_matrix;
+ * }
+ * ```
+ *
+ * @note This class inherits from std::optional<T> which implies that most
+ * of the public interface of std::optional is available.
+ */
+template <typename T>
+class Lazy : public std::optional<T>
+{
+public:
+  /**
+   * Default Constructor.
+   */
+  Lazy() = default;
+
+
+  /**
+   * Copy constructor.
+   */
+  Lazy(const Lazy &other)
+    : std::optional<T>(other)
+  {
+    //
+    // The only legal operation for a std::once_flag is to default
+    // construct it which sets its internal state to "function not
+    // called". This is OK as we first check the std::optional<T> whether
+    // it contains a value.
+    //
+  }
+
+
+  /**
+   * Move constructor.
+   */
+  Lazy(Lazy &&other) noexcept
+    : std::optional<T>(other)
+  {
+    //
+    // The only legal operation for a std::once_flag is to default
+    // construct it which sets its internal state to "function not called".
+    // This is OK as we first check the std::optional<T> whether it
+    // contains a value.
+    //
+  }
+
+
+  /**
+   * Copy assignment.
+   */
+  Lazy &
+  operator=(const Lazy &other)
+  {
+    static_cast<std::optional<T> &>(*this) =
+      static_cast<const std::optional<T> &>(other);
+
+    //
+    // Copy assignment might reset the std::optional<T> with an empty one
+    // containing no value. Thus, we have to reset the std::once_flag to
+    // "function not called".
+    //
+    flag.~once_flag();
+    new (&flag) std::once_flag;
+
+    return *this;
+  }
+
+
+  /**
+   * Move assignment.
+   */
+  Lazy &
+  operator=(Lazy &&other) noexcept
+  {
+    static_cast<std::optional<T> &>(*this) =
+      static_cast<std::optional<T> &&>(other);
+
+    //
+    // Copy assignment might reset the std::optional<T> with an empty one
+    // containing no value. Thus, we have to reset the std::once_flag to
+    // "function not called".
+    //
+    flag.~once_flag();
+    new (&flag) std::once_flag;
+
+    return *this;
+  }
+
+
+  /**
+   * Initialize the wrapped object.
+   *
+   * If the contained object is already initialized this function simply
+   * returns and does nothing.
+   *
+   * If, instead, the object has not yet been initialized then the @p
+   * creator lambda function is called to initialize the contained object.
+   * This operation is thread safe: The initialize() method ensures that
+   * the creator is only called once on one of the calling threads and that
+   * the initialization result (which is stored in the std::optional) is
+   * visible on all threads.
+   */
+  template <typename Lambda>
+  DEAL_II_ALWAYS_INLINE inline void
+  initialize(const Lambda &creator)
```

Comment:
Perhaps call this function `maybe_initialize()`?

Reply:
Or `initialize_if_necessary()`?

Reply:
I dislike `maybe_initialize`. Other options would be `initialize_if_needed()`, or - which I would prefer - `ensure_initialized()`.
Just pick one.

Reply:
`ensure_initialized()` sounds good to me.

Reply:
@bangerth On it. Will ping again when I am done with the renaming, documentation and guarding the `std::optional` modifiers.